provider/pkg/resources/cfn_custom_resource.go

```diff
@@ -0,0 +1,681 @@
+		return fmt.Errorf("failed to delete custom resource: %s", response.Reason)
+	}
+
+	glog.V(9).Infof("%s custom resource deletion succeeded", label)
+	return nil
+}
+
+// Update updates the custom resource with the given inputs and state by invoking the Lambda function with the UPDATE request type.
+// If the update is successful and the physical resource ID has changed,
+// it deletes the old resource. The function returns the updated resource properties or an error.
+//
+// Update Flow:
+//
+//   ┌─────────┐   ┌─────────────┐   ┌───────────────┐   ┌──────────┐
+//   │ Update  │   │Generate S3  │   │ Invoke Lambda │   │ Wait for │
+//   │ Request ├──►│Presigned URL├──►│ with UPDATE   ├──►│ Response │
+//   └─────────┘   └─────────────┘   │ RequestType   │   └────┬─────┘
+//                                   └───────────────┘        │
+//                                                            ▼
+//                                                    ┌───────────────┐
+//                                                    │Check Physical │
+//                                                    │Resource ID    │
+//                                                    └───────┬───────┘
+//                                                            │
+//                                                            ▼
+//                                                    ┌───────────────┐
+//                                                    │Delete Old     │
+//                                                    │Resource       │
+//                                                    │(if ID changed)│
+//                                                    └───────┬───────┘
+//                                                            │
+//                                                            ▼
+//                                                    ┌───────────────┐
+//                                                    │Return updated │
+//                                                    │Physical       │
+//                                                    │Resource ID &  │
+//                                                    │Outputs        │
+//                                                    └───────────────┘
+//
+// The Update operation:
+// 1. Generates a presigned S3 URL for response collection
+// 2. Constructs a CloudFormation UPDATE event with:
+//    - Existing PhysicalResourceId
+//    - Old and new ResourceProperties
+//    - All standard CloudFormation fields
+// 3. Invokes Lambda and collects response
+// 4. If PhysicalResourceId changes:
+//    - Initiates cleanup of old resource
+//    - Sends DELETE event for old PhysicalResourceId
+// 5. Returns updated properties and new PhysicalResourceId
+func (c *cfnCustomResource) Update(ctx context.Context, urn urn.URN, id string, inputs, oldInputs, state resource.PropertyMap, timeout time.Duration) (resource.PropertyMap, error) {
```

Comment:
Is there a way for us to get the resource options here, specifically the `retainOnDelete` option? I know there were times that I would set the `retainOnDelete` to `true` for custom resources because they didn't handle delete well.

Reply:
Sadly, not that I'm aware of: https://pulumi-developer-docs.readthedocs.io/latest/docs/_generated/proto/provider.html#pulumirpc-updaterequest

Reply:
Only MLCs have access to retainOnDelete it seems:
https://github.com/pulumi/pulumi/blob/master/proto/pulumi/provider.proto#L473

For CustomResources that are traditional I don't think the engine ever  cooperates with the provider on this. The engine handles it all engine-side.